File: myIoTGrid.Sensor/lib/sensor/src/sensor_types.cpp
#include "sensor_interface.h"
#include <unordered_map>

namespace sensor {
namespace SensorTypes {

namespace {

// Lookup table for sensor types
const std::unordered_map<std::string, const SensorTypeInfo*> sensorTypeMap = {
    {"temperature", &TEMPERATURE},
    {"humidity", &HUMIDITY},
    {"pressure", &PRESSURE},
    {"water_level", &WATER_LEVEL},
    {"co2", &CO2},
    {"pm25", &PM25},
    {"pm10", &PM10},
    {"soil_moisture", &SOIL_MOISTURE},
    {"light", &LIGHT},
    {"uv", &UV},
    {"wind_speed", &WIND_SPEED},
    {"rainfall", &RAINFALL},
    {"battery", &BATTERY},
    {"rssi", &RSSI}
};

} // anonymous namespace

const SensorTypeInfo* getInfo(const std::string& type) {
    auto it = sensorTypeMap.find(type);
    if (it != sensorTypeMap.end()) {
        return it->second;
    }
    return nullptr;
}

} // namespace SensorTypes
} // namespace sensor

File: myIoTGrid.Shared/myIoTGrid.Shared.Common/Constants/SensorTypes.cs
namespace myIoTGrid.Shared.Common.Constants;

/// <summary>
/// Standard sensor types and their units.
/// Used for validation and display purposes.
/// </summary>
public static class SensorTypes
{
    /// <summary>
    /// Mapping of sensor type codes to their standard units
    /// </summary>
    public static readonly Dictionary<string, string> Units = new()
    {
        ["temperature"] = "°C",
        ["humidity"] = "%",
        ["pressure"] = "hPa",
        ["water_level"] = "cm",
        ["flow_rate"] = "m/s",
        ["rainfall"] = "mm",
        ["co2"] = "ppm",
        ["pm25"] = "µg/m³",
        ["pm10"] = "µg/m³",
        ["pm1"] = "µg/m³",
        ["soil_moisture"] = "%",
        ["light"] = "lux",
        ["uv"] = "index",
        ["wind_speed"] = "m/s",
        ["wind_direction"] = "°",
        ["battery"] = "%",
        ["rssi"] = "dBm",
        ["voltage"] = "V",
        ["current"] = "A",
        ["power"] = "W",
        ["energy"] = "kWh",
        ["altitude"] = "m",
        ["speed"] = "km/h",
        ["distance"] = "cm",
        ["satellites"] = "",
        ["hdop"] = "",
        ["fix_type"] = ""
    };

    /// <summary>
    /// Gets the standard unit for a sensor type
    /// </summary>
    /// <param name="type">The sensor type code</param>
    /// <returns>The unit string or "unknown" if not found</returns>
    public static string GetUnit(string type)
        => Units.TryGetValue(type.ToLower(), out var unit) ? unit : "unknown";

    /// <summary>
    /// Validates if a sensor type is known
    /// </summary>
    /// <param name="type">The sensor type code to validate</param>
    /// <returns>True if this is a known sensor type</returns>
    public static bool IsValidType(string type)
        => Units.ContainsKey(type.ToLower());

    /// <summary>
    /// Gets all known sensor type codes
    /// </summary>
    public static IEnumerable<string> GetAllTypes()
        => Units.Keys;
}
